core/metacling/src/TCling.cxx

```diff
@@ -6919,14 +6943,87 @@ const char* TCling::GetClassSharedLibs(const char* cls)
    return 0;
 }
 
+/// This interface returns a list of dependent libraries in the form:
+/// lib libA.so libB.so libC.so. The first library is the library we are
+/// searching dependencies for.
+/// Note: In order to speed up the search, we display the dependencies of the
+/// libraries which are not yet loaded. For instance, if libB.so was loaded
+/// the list would contain: lib libA.so libC.so.
+static std::string GetSharedLibDepsSlow(std::string lib,
+                                        cling::Interpreter *interp,
+                                        bool skipLoadedLibs = true)
+{
+   TString LibFullPath(lib);
+   if (!llvm::sys::path::is_absolute(lib)) {
+      if (!gSystem->FindDynamicLibrary(LibFullPath, /*quiet=*/true)) {
+         Error("GetSharedLibDepsForModule", "Cannot find library '%s'", lib.c_str());
+         return "";
+      }
+   } else {
+      lib = llvm::sys::path::filename(lib);
+   }
+
+   auto ObjF = llvm::object::ObjectFile::createObjectFile(LibFullPath.Data());
+   if (!ObjF)
+      return "";
+
+   llvm::object::ObjectFile *BinObjFile = ObjF.get().getBinary();
+
+   std::set<string> DedupSet;
+   std::string Result = lib + ' ';
+   for (const auto &S : BinObjFile->symbols()) {
```

Comment:
What about the symbols of libraries that `lib` depends on? Shouldn't this function recurse over dependencies?

Reply:
It seems impossible to recurse in the generic case. Maybe if the library had explicit linking we could iterate through, but if that was the case we won't have to resolve the symbol, it'd have been resolved by the dynamic linker already.

Reply:
I don't understand any of this :-(

The interface "returns a list of dependent libraries" (from the doc). IIUC it does that even for not explicitly linked libraries (else why does it iterate through symbols?)

If one of the dependent libraries itself depends on another library - shouldn't that other library also be enumerated as a dependency of the original library?

Reply:
It is a question of intent. This PR is trying to do as much or as little to replace the rootmap file logic. In the rootmap-based implementation, rootcling would listen to a bunch of deserialized declarations and try to guess which library it came from. It did not do any transitive dependency resolution.

The proposed algorithm is much more precise but introduce O(n^3) complexity, that is for N undefined symbols, iterate over M symbols from L libraries. Because of that I had to only enable it for ACLiC where the N is presumably small. We had a reduced a little M on ELF binaries by using bloom filters. Ideally, with a lot of efforts we could further reduce the complexity.

A simple (straight forward) implementation of the suggestion would introduce higher complexity. We cannot really use simple caching techniques (such as a map) because we are tight on memory. I'd be in favor of enhancing the algorithm to resolve transitive dependencies once we optimize it to reduce the cubic complexity. Although, nowhere in the autoloading techniques (neither in the owning module-based one) we resolve transitive dependencies.

Reply:
Thanks. Can we rename the function to `GetSharedLibImmediateDepsSlow()` or something, to clarify that this is not meant to come up with the exhaustive list? Or alternatively clarify this in the doc?

Now - rootmap files tell `TCling` which library to load and which header to parse for a given class. I don't remember nor see that we store dependent libraries in there. Can you remind me of an example?

Reply:
Renamed. The code doing library autoloading is [here](https://github.com/root-project/root/blob/master/core/metacling/src/TCling.cxx#L6771). We get the owning module of a used decl and get the relevant library from the modulemap file.

Reply:
Thanks - ok that's mostly used by ACLiC. It notes which libs are currently loaded and will load them before loading the ACLiC'ed lib.

The current setup notes *all* libs. I.e. even if libA depends on libB depends on libC - libC will be noted as "must load". IIUC your replacement code will only load libs that resolve symbols *of `lib`*, but not *of those that lib depends on*.

That's not an issue for explicit linking, which is what we do on Linux and Mac. It is likely going to be an issue if that's not happening (rare or Windows).

Can you add a comment along these lines? I don't believe that this missing feature should block the PR.

Reply:
I do not understand this part `, but not of those that lib depends on.`

We essentially disable explicit linking when it comes to ACLiC with modules disabled and fall back to symbol resolution if that's what you meant. The same symbol resolution mechanisms for those undefined symbols will apply to windows. We do not rely on hooks from dyld (although that would bring the algorithm complexity down), so I do not see how windows can be affected. Could you elaborate?